src/Shape.cpp

```diff
@@ -22,6 +22,80 @@ bool Shape::intersect(const Ray &, float t0, float t1, double& distance) const {
     return false;
 }
 
+
+/** @brief Shape::intersect() determines whether a point p lies within the shape.
+ *  @return bool True means point p lies inside the shape.
+ *  @math Let the line segment P connect points p and an arbitrary point p_out outside of the shape
+ *  We count the number of intersections between P and the shape. A positive number means point p is inside the shape.
+ *  We use plucker coordinates to determine whether or not a triangle intersects line segment P
+ **/
+
+bool Shape::intersect(const Vector3& p) const {
+    if (p.length2() > mesh_.getSquaredMaxRadius()){
+        return false;
+    }
+
+    int intersect_count = 0;
+
+    // determine plucker coordinates of line p
+    Vector3 p_out = Vector3(1.1 * mesh_.getMaxRadius(), 0, 0);
+    Vector3 p_U = p - p_out;
+    Vector3 p_V = p.cross(p_out);
+
+    // load triangles
+    const std::vector<geo::Vector3>& t_points = mesh_.getPoints();
+    const std::vector<TriangleI>& triangles_i = mesh_.getTriangleIs();
+    for(std::vector<TriangleI>::const_iterator it = triangles_i.begin(); it != triangles_i.end(); ++it) {
+        Vector3 v1 = t_points[it->i1_];
+        Vector3 v2 = t_points[it->i2_];
+        Vector3 v3 = t_points[it->i3_];
+
+        Vector3 e1_U = v1 - v2;
+        Vector3 e2_U = v2 - v3;
+        Vector3 e3_U = v3 - v1;
+
+        Vector3 e1_V = v1.cross(v2);
+        Vector3 e2_V = v2.cross(v3);
+        Vector3 e3_V = v3.cross(v1);
+
+        double S1 = side_operator(p_U, p_V, e1_U, e1_V);
+        double S2 = side_operator(p_U, p_V, e2_U, e2_V);
+        double S3 = side_operator(p_U, p_V, e3_U, e3_V);
+
+        if ((S1 < 0 && S2 < 0 && S3 < 0) || (S1 > 0 && S2 > 0 && S3 > 0)) { // the line passes through the triangle. now check the line segment
+            Vector3 l1_U = p_out - v1;
+            Vector3 l2_U = v1 - p;
+
+            Vector3 l1_V = p_out.cross(v1);
+            Vector3 l2_V = v1.cross(p);
+
+            double S4 = side_operator(l1_U, l1_V, e2_U, e2_V);
+            double S5 = side_operator(l2_U, l2_V, e2_U, e2_V);
+
+            if ((S4 > 0 && S5 < 0) || (S4 < 0 && S5 > 0)) {
+                if (S1 < 0 && S2 < 0 && S3 < 0) {
+                    intersect_count--;
+                }
+                else if(S1 > 0 && S2 > 0 && S3 > 0) {
+                    intersect_count++;
+                }
+            }
+        }
+    }
+
+   if (intersect_count < 0 || intersect_count > 1) {
+        std::cout << "intersect_count is " << intersect_count << ", it should be 0 or 1!" << std::endl;
+        return false;
+    }
+
+    return intersect_count > 0;
+}
+
+double Shape::side_operator(Vector3& p_U, Vector3& p_V, Vector3& q_U, Vector3& q_V){
```

Comment:
It doesn't use anything in the `Shape` class, why not make it a simple function?

Its name has no meaning to me, so it's either wrong or it needs explanation or some reference that explains what you're doing.

Reply:
Depending on the c++ version that we use, you may want to declare this function as `inline`, although likely the compiler does this anyway regardless what you declare.